network/p2p/node_sampler.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+var (
+	_ NodeSampler = (*Validators)(nil)
+
+	_ validators.Connector = (*Peers)(nil)
+	_ NodeSampler          = (*Peers)(nil)
+)
+
+type NodeSampler interface {
+	// Sample returns at most [n] nodes. This may return fewer nodes if fewer
+	// than [n] are available.
+	Sample(n int) ([]ids.NodeID, error)
+}
+
+type Validators struct {
+	SubnetID   ids.ID
+	Validators validators.State
+}
+
+func (v Validators) Sample(n int) ([]ids.NodeID, error) {
+	currentHeight, err := v.Validators.GetCurrentHeight(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current height: %w", err)
+	}
+
+	validatorSet, err := v.Validators.GetValidatorSet(context.TODO(), currentHeight, v.SubnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current validator set: %w", err)
+	}
```

Comment:
This should be cached - we don't need exact validator sets - so maybe update it every few minutes max